applications/TopologyOptimizationApplication/custom_elements/small_displacement_simp_element.cpp

```diff
@@ -0,0 +1,286 @@
+        const double E_initial = this->GetValue(YOUNGS_MODULUS_0);
+        const double E_current = this->GetValue(YOUNG_MODULUS);
+        const double penalty   = this->GetValue(PENAL);
+        const double x_phys    = this->GetValue(X_PHYS);
+
+        // Get element stiffness matrix and modify it with the factor that considers the adjusted Youngs Modulus according to the SIMP method
+        // Note that Ke0 is computed based on the originally provided Youngs Modulus in the .mdpa-file
+        MatrixType Ke0 = Matrix();
+        this->CalculateLeftHandSide(Ke0, const_cast <ProcessInfo&>(rCurrentProcessInfo));
+        double E_new     = (E_min + pow(x_phys, penalty) * (E_initial - E_min));
+
+        ///Normalize the youngs modulus
+        double factor    = E_new/E_current; 
+        MatrixType Ke = Ke0 * factor;
+
+        // Loop through nodes of elements and create elemental displacement vector "ue"
+        Element::GeometryType& rGeom = this->GetGeometry();
+        unsigned int NumNodes = rGeom.PointsNumber(); 
+
+        // Resize "ue" according to element type
+        Vector ue;
+        ue.resize(NumNodes * 3);
+
+        // Set the displacement obtained from the FE-Analysis
+        for (unsigned int node_i = 0; node_i < NumNodes; node_i++) {
+            array_1d<double, 3> &CurrentDisplacement = rGeom[node_i].FastGetSolutionStepValue(DISPLACEMENT);
+            ue[3 * node_i + 0] = CurrentDisplacement[0];
+            ue[3 * node_i + 1] = CurrentDisplacement[1];
+            ue[3 * node_i + 2] = CurrentDisplacement[2];
+        }
+
+        // Calculate trans(ue)*Ke0*ue
+        Vector intermediateVector;
+        intermediateVector.resize(NumNodes * 3);
+        intermediateVector = prod(trans(ue), Ke0);
+        double ue_Ke0_ue = inner_prod(intermediateVector, ue);
+
+        if (rVariable == DCDX)
+        {
+            // Calculation of the compliance sensitivities DCDX
+            double dcdx = (-penalty)* (E_initial - E_min) * pow(x_phys, penalty - 1) * ue_Ke0_ue;
+            this->SetValue(DCDX, dcdx); 
+        }
+        if (rVariable == LOCAL_STRAIN_ENERGY)
+        {
+            // Calculation of the local strain energy (requires Ke)
+            double local_strain_energy = factor * ue_Ke0_ue;
+            this->SetValue(LOCAL_STRAIN_ENERGY, local_strain_energy);
+
+        }
+
+    } else if (rVariable == DVDX) {
+		// Calculation of the volume sensitivities DVDX
+        double element_size = this->GetValue(INITIAL_ELEMENT_SIZE);
+        this->SetValue(DVDX, element_size*1);
+	}
+
+    KRATOS_CATCH( "" )
+} 
+ 
+// =============================================================================================================================================
+// =============================================================================================================================================
+
+void SmallDisplacementSIMPElement::SetElementData(const KinematicVariables& rThisKinematicVariables,
```

Comment:
What's the difference with the base class?

Reply:
SetElementData was implemented in the original TopOpt tool and was therefore based on SolidMechanicsApplication. Is there a function in the base class I could replace it with?

Reply:
Yes, https://github.com/KratosMultiphysics/Kratos/blob/f55a2c57ac8102038bdb46428cb2d2a95c7b92cc/applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp#L1473

Derived in the corresponding elements